### ISSUE TITLE
UninitializedFieldError thrown for default initialization 

### ISSUE THREAD

**Issue by steinybot** [scala/bug#10772]
The bug that was initially reported in https://github.com/scala/bug/issues/10437 does not seem to be completely fixed in 2.12.4.

I encountered this in the same way that was reported in https://github.com/akka/akka/issues/23453.

Here is how to reproduce it:

`build.sbt`
```scala
scalaVersion := "2.12.4"
scalacOptions += "-Xcheckinit"
libraryDependencies += "com.typesafe.akka" %% "akka-stream" % "2.5.11"
```

`src/main/scala/Foo.scala`
```scala
import akka.actor.ActorSystem
import akka.event.Logging
import akka.stream._
import akka.stream.scaladsl._
import akka.stream.stage._

object Foo extends App {

  val system = ActorSystem()
  implicit val materializer = ActorMaterializer.create(system)

  val source = Source.single("foo")
  val log = new GraphStage[FlowShape[String, String]] {
    val in = Inlet[String]("Log.in")
    val out = Outlet[String]("Log.out")
    override val shape = FlowShape(in, out)
    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = {
      new GraphStageLogic(shape) with StageLogging with OutHandler with InHandler {
        override def onPush(): Unit = {
          val elem = grab(in)
          log.log(Logging.InfoLevel, "logging some stuff...")
          push(out, elem)
        }
        override def onPull(): Unit = pull(in)
        setHandlers(in, out, this)
      }
    }
  }
  source.via(log).runWith(Sink.ignore)
}
```

```
→ sbt run
[info] Loading settings from idea.sbt ...
[info] Loading global plugins from /Users/jason/.sbt/1.0/plugins
[info] Loading project definition from /source/scalac-checkinit-bug/project
[info] Loading settings from build.sbt ...
[info] Set current project to scalac-checkinit-bug (in build file:/source/scalac-checkinit-bug/)
[info] Running Foo
[ERROR] [03/13/2018 14:52:43.773] [default-akka.actor.default-dispatcher-3] [akka://default/system/StreamSupervisor-0/flow-0-0-ignoreSink] Error in stage [Foo$$anon$2@461d6c93]: Uninitialized field: /source/scalac-checkinit-bug/src/main/scala/Foo.scala: 18
scala.UninitializedFieldError: Uninitialized field: /source/scalac-checkinit-bug/src/main/scala/Foo.scala: 18
	at Foo$$anon$2$$anon$1.akka$stream$stage$StageLogging$$_log(Foo.scala:18)
	at akka.stream.stage.StageLogging.log(StageLogging.scala:29)
	at akka.stream.stage.StageLogging.log$(StageLogging.scala:27)
	at Foo$$anon$2$$anon$1.log(Foo.scala:18)
	at Foo$$anon$2$$anon$1.onPush(Foo.scala:21)
	at akka.stream.impl.fusing.GraphInterpreter.processPush(GraphInterpreter.scala:519)
	at akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:508)
	at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:378)
	at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:585)
	at akka.stream.impl.fusing.GraphInterpreterShell.init(ActorGraphInterpreter.scala:555)
	at akka.stream.impl.fusing.ActorGraphInterpreter.tryInit(ActorGraphInterpreter.scala:676)
	at akka.stream.impl.fusing.ActorGraphInterpreter.preStart(ActorGraphInterpreter.scala:724)
	at akka.actor.Actor.aroundPreStart(Actor.scala:528)
	at akka.actor.Actor.aroundPreStart$(Actor.scala:528)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundPreStart(ActorGraphInterpreter.scala:667)
	at akka.actor.ActorCell.create(ActorCell.scala:654)
	at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:525)
	at akka.actor.ActorCell.systemInvoke(ActorCell.scala:547)
	at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:282)
	at akka.dispatch.Mailbox.run(Mailbox.scala:223)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

Apologies that this is isn't as minimal as it probably could be. I tried a few trivial cases but I must have been missing something (likely the same thing that what was missing from the original fix) because I could not reproduce it.

**Comment by ngbinh** [comment]
isn't `log` in `log.log` is shadowed by previous `val log = new GraphStage[FlowShape[String, String]] `? If it is then there is a cyclic dependency here.

**Comment by lrytz** [comment]
This is actually a duplicate of https://github.com/scala/bug/issues/10692, fixed in 2.12.5
